pkg/agent/packetcapture/packetcapture_controller_test.go

```diff
@@ -0,0 +1,597 @@
+	}
+	var objs []runtime.Object
+	for i := 0; i < 20; i++ {
+		objs = append(objs, genTestCR(nameFunc(i), int32(testCaptureNum)))
+	}
+	pcc := newFakePacketCaptureController(t, nil, objs)
+	pcc.sftpUploader = &testUploader{url: testFTPUrl}
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	pcc.crdInformerFactory.Start(stopCh)
+	pcc.crdInformerFactory.WaitForCacheSync(stopCh)
+	pcc.informerFactory.Start(stopCh)
+	pcc.informerFactory.WaitForCacheSync(stopCh)
+	go pcc.Run(stopCh)
+	assert.Eventually(t, func() bool {
+		items, err := pcc.crdClient.CrdV1alpha1().PacketCaptures().List(context.Background(), metav1.ListOptions{})
+		if err != nil {
+			return false
+		}
+		for _, result := range items.Items {
+			for _, cond := range result.Status.Conditions {
+				if cond.Type == crdv1alpha1.PacketCaptureComplete || cond.Type == crdv1alpha1.PacketCaptureFileUploaded {
+					if cond.Status == metav1.ConditionFalse {
+						return false
+					}
+				}
+			}
+		}
+		pcc.mutex.Lock()
+		if pcc.numRunningCaptures != 0 {
+			return false
+		}
+		pcc.mutex.Unlock()
+		return true
+	}, 5*time.Second, 50*time.Millisecond)
+	for i := 0; i < 20; i++ {
+		err := pcc.crdClient.CrdV1alpha1().PacketCaptures().Delete(context.TODO(), nameFunc(i), metav1.DeleteOptions{})
+		require.NoError(t, err)
+	}
+	assert.Eventually(t, func() bool {
+		pcc.mutex.Lock()
+		if len(pcc.captures) != 0 {
+			return false
+		}
+		pcc.mutex.Unlock()
+		return true
+	}, 2*time.Second, 20*time.Millisecond)
+
+}
+
+// TestPacketCaptureControllerRun was used to validate the whole run process is working. It doesn't wait for
+// the testing pc to finish. on sandbox env, no good solution to open raw socket.
+func TestPacketCaptureControllerRun(t *testing.T) {
+	pcs := []struct {
+		name                 string
+		pc                   *crdv1alpha1.PacketCapture
+		expectCompleteStatus metav1.ConditionStatus
+		expectUploadStatus   metav1.ConditionStatus
+	}{
+		{
+			name:                 "start packetcapture",
+			expectCompleteStatus: metav1.ConditionTrue,
+			expectUploadStatus:   metav1.ConditionTrue,
+			pc: &crdv1alpha1.PacketCapture{
+				ObjectMeta: metav1.ObjectMeta{Name: "pc1", UID: "uid1"},
+				Spec: crdv1alpha1.PacketCaptureSpec{
+					Source: crdv1alpha1.Source{
+						Pod: &crdv1alpha1.PodReference{
+							Namespace: pod1.Namespace,
+							Name:      pod1.Name,
+						},
+					},
+					Destination: crdv1alpha1.Destination{
+						Pod: &crdv1alpha1.PodReference{
+							Namespace: pod2.Namespace,
+							Name:      pod2.Name,
+						},
+					},
+					CaptureConfig: crdv1alpha1.CaptureConfig{
+						FirstN: &crdv1alpha1.PacketCaptureFirstNConfig{
+							Number: 15,
+						},
+					},
+					Packet: &crdv1alpha1.Packet{
+						Protocol: &icmpProto,
+					},
+					FileServer: &crdv1alpha1.PacketCaptureFileServer{
+						URL: "sftp://127.0.0.1:22/aaa",
+					},
+					Timeout: &testCaptureTimeout,
+				},
+			},
+		},
+		{
+			name:                 "parse ip",
```

Comment:
wouldn't it make sense to unit test `parseIPs` in isolation, assuming this is the purpose of this test case?

Reply:
sure it does. Is this not the best practices? I thought let the tests going through all the run process would be better and it didn't cause much overhead.